File: Modules/SEDO/EMS/SIE.Web.EMS/SpareParts/Commands/CopySparePartStoreCommand.cs
using SIE.EMS.SpareParts;
using SIE.Web.Command;
using System;

namespace SIE.Web.EMS.SpareParts.Commands
{
    /// <summary>
    /// 备件入库复制命令
    /// </summary>
    [JsCommand("SIE.Web.EMS.SpareParts.Commands.CopySparePartStoreCommand")]
    public class CopySparePartStoreCommand : ViewCommand
    {
        /// <summary>
        /// 执行复制操作
        /// </summary>
        /// <param name="args">args</param>
        /// <param name="scope">scope</param>
        /// <returns>备件入库单号</returns>
        protected override object Excute(ViewArgs args, string scope)
        {
            var para = args.Data.ToJsonObject<Para>();
            if (para.IsStoreCode)
            {
                return RT.Service.Resolve<SparePartController>().GetStoreCode();
            }
            else
            {
                return RT.Service.Resolve<SparePartController>().GetDetailCodes(para.DtlCount);
            }

        }
    }

    /// <summary>
    /// 参数
    /// </summary>
    [Serializable]
    public class Para
    {
        /// <summary>
        /// 是否入库单编码
        /// </summary>
        public bool IsStoreCode { get; set; }

        /// <summary>
        /// 明细数量
        /// </summary>
        public int DtlCount { get; set; }
    }
}

File: Modules/SEDO/EMS/SIE.Web.EMS/SpareParts/Commands/CopySparePartStoreCommand.js
SIE.defineCommand('SIE.Web.EMS.SpareParts.Commands.CopySparePartStoreCommand', {
    extend: 'SIE.cmd.Copy',
    meta: { text: "复制新增", group: "edit", splitTo: "添加", iconCls: "icon-AddEntity icon-green" },
    onEditting: function (entity) {
        var me = this;
        if (entity) {
            var para = {
                IsStoreCode: true,
                DtlCount: 0
            }

            //生成单号
            me.view.execute({
                data: para,
                success: function (res) {
                    entity.setStoreCode(res.Result);
                    entity.setStoreStatus(0);

                    //子表生成新序列或，清空孙表数据
                    var dtllist = entity.StoreDetailList().data;
                    if (dtllist.items.length > 0) {

                        //生成批次号
                        para.IsStoreCode = false;
                        para.DtlCount = dtllist.items.length; 
                        me.view.execute({
                            data: para,
                            success: function (res) {

                                var batchNos = res.Result;
                                for (i = 0; i < dtllist.items.length; i++) {
                                    dtllist.items[i].OrderNumberList().data.items.removeAll();
                                    dtllist.items[i].setBatchNumber(batchNos[i]);
                                    dtllist.items[i].setStoreStatus(0);
                                };
                            }
                        }, me.view);


                    };
                }
            }, me.view);
        }
    }
});